### ISSUE TITLE
DAQInterface appears to leak file descriptors

### ISSUE THREAD

**Issue by eflumerf** [art-daq/artdaq_daqinterface#88]
*This issue has been migrated from https://cdcvs.fnal.gov/redmine/issues/20757 (FNAL account required)*
*Originally created by @jcfreeman2 on 2018-09-06 16:51:02*

----

From Eric:

"DAQInterface appears to leak file descriptors (possibly handles to per-run log files?)...I did 297 runs in one instance, and it failed to boot after due to a "too many open files" error. "

**Comment by eflumerf** [comment]
*Comment by @eflumerf on 2018-09-07 12:43:05*

----

I made another long-running instance, and there's a large number of PIPEs, apparently between two threads within DAQInterface. I have captured the output of lsof into a log file, as well as the output of ls -l /proc/<PID>/fd.

**Comment by eflumerf** [comment]
*Comment by @eflumerf on 2018-09-07 20:51:26*

----

After a bit more testing, it appears that the leak happens on Terminate-Boot cycles, not Stop-Start cycles.

**Comment by eflumerf** [comment]
*Comment by @jcfreeman2 on 2018-09-16 01:19:03*

----

Solved with commit e05a76e400de3ba1382f8424c9cf2f333179d717. I'll just quote the commit comments:

JCF: The cut-and-pasted deepsuppression context manager I got from stackoverflow.com back in 2014 had
a bug, in that it wasn't closing all file descriptors in deepsuppression.*_exit_*.  Whenever deepsuppression() was called by DAQInterface, it was leaking file descriptors, a problem observed by Eric in Redmine Issue 20757.  This commit properly closes the file 
descriptors.
